flax/linen/module.py

```diff
@@ -547,12 +576,59 @@ def setup(self):
 
                # Accessing `submodule.variables` is now safe.
 
-      3. Immediately when a module is constructed inside a method wrapped with 
+      3. Immediately when a module is constructed inside a method wrapped with
          :meth:`compact`.
 
     """
     pass
 
+  @wrap_method
+  def _adopt_toplevel_modules(self):
```

Comment:
An easier solution would be to a "bind-to-first" behaviour. Currently we fail if you bind a module twice but we could just bind the parent on the first attribute assignment and make later bindings a no-op. This would be more consistent because you don't need the top-level exception case to be handled separately. Instead you would always bind modules passed as attributes which were unbound before. This also removes the distinction between constructor attributes and setup attributes. What do you think?

Reply:
The only thing this doesn't address is sharing-by-reference semantics.  Basically if you write code w. shared refs at the top, then we don't respect parameter-sharing relationships in the top-init'd module in a simple bind-to-first implementation.  Maybe we say that's just how it is, but it _is_ a different set of rules for "top-level style" vs "internal style" which seems like something painful to explain or debug later.  A subset of our users coming from pytorch really seem to like defining models in this "top-level style", and I'm sure someone eventually will try setting up sharing and wonder why it's broken.

Reply:
I should also add that "bind to first" doesn't save a ton of code, as you can't bind directly in `__setattr__` since the parent module hasn't been instantiated yet, you have to do this at the end of post-init in a helper function in any case.

Reply:
I still don't get how top-level is a special case:

I would expect the following to work after with PR:
```
class Foo(nn.Module):
  def setup(self):
    self.foo = Bar(bar=nn.Dense(1))

class Bar(nn.Module):
  bar: Any

  def setup(self):
    pass

  def __call__(self, x):
    return self.bar(x)

x = jnp.ones((2,))
Foo().init(random.PRNGKey(0), x)
```
but it actually raises an error because it will not bind to `bar.bar` as I would expect. It would be weird if this same pattern did work if you pass a Module instance to the top-level constructor right (which to me makes less sense)?